src/query/api/v1/handler/prom/read.go

```diff
@@ -0,0 +1,179 @@
+// Package prom provides custom handlers that support the prometheus
+// query endpoints.
+package prom
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/m3db/m3/src/query/api/v1/options"
+	"github.com/m3db/m3/src/query/storage/prometheus"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/promql"
+	promstorage "github.com/prometheus/prometheus/storage"
+	"github.com/uber-go/tally"
+	"go.uber.org/zap"
+)
+
+type readHandler struct {
+	engine    *promql.Engine
+	queryable promstorage.Queryable
+	hOpts     options.HandlerOptions
+	scope     tally.Scope
+	logger    *zap.Logger
+}
+
+type readRequest struct {
+	query         string
+	start, end    time.Time
+	step, timeout time.Duration
+}
+
+func newReadHandler(
+	opts Options,
+	hOpts options.HandlerOptions,
+	queryable promstorage.Queryable,
+) http.Handler {
+	scope := hOpts.InstrumentOpts().MetricsScope().Tagged(
+		map[string]string{"handler": "prometheus-read"},
+	)
+	return &readHandler{
+		engine:    opts.PromQLEngine,
+		queryable: queryable,
+		hOpts:     hOpts,
+		scope:     scope,
+		logger:    hOpts.InstrumentOpts().Logger(),
+	}
+}
+
+func (h *readHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	request, err := parseRequest(r)
+	if err != nil {
+		respondError(w, err, http.StatusBadRequest)
+		return
+	}
+
+	fetchOptions, fetchErr := h.hOpts.FetchOptionsBuilder().NewFetchOptions(r)
```

Comment:
FetchOptionsBuilder does not need to come from handler options as it can be available on the handler when it gets created itself.